src/libsyntax/parse/parser.rs

```diff
@@ -3527,6 +3526,22 @@ impl<'a> Parser<'a> {
         }
     }
 
+    // helper function to decide whether to parse as ident binding or to try to do
+    // something more complex like range patterns
+    fn parse_as_ident(&mut self) -> bool {
+        self.look_ahead(1, |t| match *t {
+            token::OpenDelim(token::Paren) | token::OpenDelim(token::Brace) |
+            token::DotDotDot | token::ModSep | token::Not => Some(false),
+            // ensure slice patterns [a, b.., c] and [a, b, c..] don't go into the
+            // range pattern branch
+            token::DotDot => None,
```

Comment:
[Syntactic conflict](https://github.com/rust-lang/rust/issues/23121#issuecomment-154138818) between subslice patterns and exclusive range patterns is finally revealing itself.